### DISCUSSION TITLE
Proper handling of None in WHERE condition

### DISCUSSION THREAD

**Discussion by mmlynarik** [fastapi/sqlmodel#1515] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the SQLModel documentation, with the integrated search.
- [X] I already searched in Google "How to X in SQLModel" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to SQLModel but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to SQLModel but to [SQLAlchemy](https://github.com/sqlalchemy/sqlalchemy).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Optional

from sqlmodel import Field, Session, SQLModel, create_engine, select, col


class Hero(SQLModel, table=True):
    id: Optional[int] = Field(default=None, primary_key=True)
    name: str
    secret_name: str
    age: Optional[int] = None


sqlite_file_name = "database.db"
sqlite_url = f"sqlite:///{sqlite_file_name}"

engine = create_engine(sqlite_url, echo=True)


def create_db_and_tables():
    SQLModel.metadata.create_all(engine)


def create_heroes():
    hero_1 = Hero(name="Deadpond", secret_name="Dive Wilson")
    hero_2 = Hero(name="Spider-Boy", secret_name="Pedro Parqueador")
    hero_3 = Hero(name="Rusty-Man", secret_name="Tommy Sharp", age=48)

    with Session(engine) as session:
        session.add_all([hero_1, hero_2, hero_3])
        session.commit()


def select_heroes():
    with Session(engine) as session:
        statement = select(Hero).where(col(Hero.age) is None, Hero.name.like('%Deadpond%'))
        results = session.exec(statement)
        for hero in results:
            print(hero)
```

### Description

To make the where condition `Hero.age is None` functional, because currently it is evaluated as false even though it should be true. Otherwise, one needs to use `Hero.age == None`, which raises linting error.

### Wanted Solution

To make the where condition `Hero.age is None` functional, because currently it is evaluated as false even though it should be true. Otherwise, one needs to use `Hero.age == None`, which raises linting error.

### Wanted Code

```python
from typing import Optional

from sqlmodel import Field, Session, SQLModel, create_engine, select, col


class Hero(SQLModel, table=True):
    id: Optional[int] = Field(default=None, primary_key=True)
    name: str
    secret_name: str
    age: Optional[int] = None


sqlite_file_name = "database.db"
sqlite_url = f"sqlite:///{sqlite_file_name}"

engine = create_engine(sqlite_url, echo=True)


def create_db_and_tables():
    SQLModel.metadata.create_all(engine)


def create_heroes():
    hero_1 = Hero(name="Deadpond", secret_name="Dive Wilson")
    hero_2 = Hero(name="Spider-Boy", secret_name="Pedro Parqueador")
    hero_3 = Hero(name="Rusty-Man", secret_name="Tommy Sharp", age=48)

    with Session(engine) as session:
        session.add_all([hero_1, hero_2, hero_3])
        session.commit()


def select_heroes():
    with Session(engine) as session:
        statement = select(Hero).where(col(Hero.age) is None, Hero.name.like('%Deadpond%'))
        results = session.exec(statement)
        for hero in results:
            print(hero)
```

### Operating System

Linux

### SQLModel Version

0.0.4

### Python Version

3.8.5

**Comment by yasamoka** [comment]
I believe this is concerned entirely with SQLAlchemy, not with SQLModel, and has to do with the required semantics to construct a `BinaryExpression` object.

`Hero.age == None` evaluates to a `BinaryExpression` object which is eventually used to construct the SQL query that the SQLAlchemy engine issues to your DBMS.

`Hero.age is None` evaluates to `False` immediately, and not a `BinaryExpression`, which short-circuits the query no matter the value of `age` in a row.

From a cursory search, it does not seem that the `is` operator can be overridden in Python. This could help explain why the only possibility is by using the `==` operator, which can be overridden.

**Comment by s-c-p** [comment]
@mmlynarik Along the lines of `.like()`, would it be acceptable if we implement something like `Hero.age.is(None)`?

**Comment by mmlynarik** [comment]
> @mmlynarik Along the lines of `.like()`, would it be acceptable if we implement something like `Hero.age.is(None)`?

If it worked as expected, then definitely yes:)

**Comment by s-c-p** [comment]
Ok, I am working on a PR implementing this, repo owners please give me some time 👍

**Comment by tc-imba** [comment, marked as answer]
You can use `Hero.age.is_(None)`

**Comment by northtree** [comment]
@tc-imba where to import `is_`?

**Comment by yasamoka** [comment]
> @tc-imba where to import `is_`?

`is_` is an attribute. You don't need to import it to use it.

**Comment by NMertsch** [comment]
My understanding is this:

`where(Hero.age is None)` cannot be implemented, because the `is` operator cannot be overloaded (`==` can be overloaded via the `__eq__` method, etc.).

`where(Hero.age.is(None))` cannot be implemented, because `is` is a reserved keyword and therefore no valid identifier.

`where(Hero.age.is_(None))` works and probably is the closest we can get. It follows [PEP 8 Naming Styles](https://peps.python.org/pep-0008/#descriptive-naming-styles) by using a trailing underscore to "avoid conflicts with Python keyword".

But `is_` is not working with autocompletion, at least not in Pycharm. `Hero.age` is inferred to be `int | None`, which is exactly what I want in 99% of cases. But it means that `Hero.age.is_` raises a warning "unresolved reference".
Not sure there is a good way to deal with it.

**Comment by northtree** [comment]
For `IS NOT NULL`, you could use `is_not`
> where(Hero.age.is_not(None))

**Reply by AdityaMayukhSom** [reply]
`Instance of 'FieldInfo' has no 'is_' member` getting this error, any suggestions?

**Comment by Cielquan** [comment]
> My understanding is this:
> 
> `where(Hero.age is None)` cannot be implemented, because the `is` operator cannot be overloaded (`==` can be overloaded via the `__eq__` method, etc.).
> 
> `where(Hero.age.is(None))` cannot be implemented, because `is` is a reserved keyword and therefore no valid identifier.
> 
> `where(Hero.age.is_(None))` works and probably is the closest we can get. It follows [PEP 8 Naming Styles](https://peps.python.org/pep-0008/#descriptive-naming-styles) by using a trailing underscore to "avoid conflicts with Python keyword".
> 
> But `is_` is not working with autocompletion, at least not in Pycharm. `Hero.age` is inferred to be `int | None`, which is exactly what I want in 99% of cases. But it means that `Hero.age.is_` raises a warning "unresolved reference". Not sure there is a good way to deal with it.

The same goes for the `in_` operator method (an presumably all the other operators as well):
```python
select(Ticket).where(Ticket.stage_id.in_(stage_ids))
```
This also let mypy and other type using tool cry out loud like `"int" has no attribute "in_"  [attr-defined]mypy(error)`

**Comment by Cielquan** [comment]
I guess the issue title could be updated to smth like `Proper handling of column operator methods in WHERE condition`

**Comment by patrickwasp** [comment]
looking for a team without heros like

`select(Team).where(Team.heros == None)`

isn't compatible with [E711](https://www.flake8rules.com/rules/E711.html)

but `select(Team).where(Team.heros is None)`

doesn't work

**Comment by NMertsch** [comment]
`Team.heros is None` cannot be made to work because the Python language does not allow overloading `is`, but does allow overloading `==`.

`Team.heros == None` checks for equality, not identity. It seems to do what you want, but it is a [PEP 8 violation](https://peps.python.org/pep-0008/#programming-recommendations).

`Team.heros.is_(None)` should also do what you want, and is more idiomatic because it checks for identity (like `is`), not equality (like `==`), following the PEP 8 recommendation.

In the end, write the code that works for you. PEP 8 is just a recommendation, and flake8 is just a tool. If you prefer to ignore a recommendation, [PEP 8 has a section about that as well](https://peps.python.org/pep-0008/#a-foolish-consistency-is-the-hobgoblin-of-little-minds).

**Comment by patrickwasp** [comment]
Besides using `# type: ignore` is there an easy way to make pylance happy? `Attribute "is_" is unknown Pylance`

**Comment by gerrior** [comment]
@patrickwasp this issue is still open. It hasn't be merged to the code base yet. 

I recommend the following code: 
```
select(Team).where(
            Team.heros == None,  # noqa E711 Comparison to `None` should be `cond is None`.
)
```

```
select(Team).where(
            Team.heros is None,  # This does not work as intended.
)
```

**Comment by ghost** [comment]
```
select(Team).where(
            Team.heros == None,  # noqa E711 Comparison to `None` should be `cond is None`.
)
```

this is working on me but: Comparison to `None` should be `cond is not None`Ruff[E711](https://docs.astral.sh/ruff/rules/none-comparison)

**Comment by Nikola-Milovic** [comment]
Why aren't there helper functions like `is_(age.Hero, 10)`? That seems like it would do the job?

**Comment by Ocyss** [comment]
I ran into this exact issue today. After reading through the various workarounds, I’m currently in a bit of an awkward spot:

- Using `.is_` (or `.is_not`) works with SQLModel/SQLAlchemy, but it completely loses type-checker autocompletion and hints in most editors.  
- Using `== None` (or `!= None`) keeps proper type hints and autocompletion, but triggers Ruff’s “comparison with None” rule `Ruff(E711)`.

For now I’ve just disabled the RUF011 rule in my project,

I’m really looking forward to Ruff’s upcoming plugin system (https://github.com/astral-sh/ruff/issues/283). Once that lands, I plan to write a small plugin that:

1. Suppresses the “comparison with None” warning specifically inside `session.exec(statement.where(...))` contexts (and similar SQLModel/SQLAlchemy query builders).  
2. Optionally auto-fixes `is None` → `== None` (and `is not None` → `!= None`) inside those contexts, so we get both correct runtime behavior and clean linting without sacrificing type hints.
